pkg/reconciler/inmemorychannel/dispatcher/inmemorychannel.go

```diff
@@ -19,87 +19,96 @@ package dispatcher
 	"knative.dev/eventing/pkg/channel"
 	"knative.dev/eventing/pkg/channel/fanout"
 	"knative.dev/eventing/pkg/channel/multichannelfanout"
-	listers "knative.dev/eventing/pkg/client/listers/messaging/v1"
-	"knative.dev/eventing/pkg/inmemorychannel"
 )
 
 // Reconciler reconciles InMemory Channels.
 type Reconciler struct {
 	eventDispatcherConfigStore *channel.EventDispatcherConfigStore
-	dispatcher                 inmemorychannel.MessageDispatcher
-	inmemorychannelLister      listers.InMemoryChannelLister
-	inmemorychannelInformer    cache.SharedIndexInformer
+	multiChannelMessageHandler multichannelfanout.MultiChannelMessageHandler
+	reporter                   channel.StatsReporter
 }
 
 func (r *Reconciler) ReconcileKind(ctx context.Context, imc *v1.InMemoryChannel) reconciler.Event {
-	// This is a special Reconciler that does the following:
-	// 1. Lists the inmemory channels.
-	// 2. Creates a multi-channel-fanout-config.
-	// 3. Calls the inmemory channel dispatcher's updateConfig func with the new multi-channel-fanout-config.
-	channels, err := r.inmemorychannelLister.List(labels.Everything())
-	if err != nil {
-		logging.FromContext(ctx).Error("Error listing InMemory channels")
-		return err
-	}
+	logging.FromContext(ctx).Infow("Reconciling", zap.Any("InMemoryChannel", imc))
 
-	inmemoryChannels := make([]*v1.InMemoryChannel, 0)
-	for _, imc := range channels {
-		if imc.Status.IsReady() {
-			inmemoryChannels = append(inmemoryChannels, imc)
-		}
+	if !imc.Status.IsReady() {
+		logging.FromContext(ctx).Debug("IMC is not ready, skipping")
+		return nil
 	}
 
-	config, err := r.newConfigFromInMemoryChannels(inmemoryChannels)
+	config, err := r.newConfigForInMemoryChannel(imc)
 	if err != nil {
-		logging.FromContext(ctx).Error("Error creating config from in memory channels")
-		return err
-	}
-	err = r.dispatcher.UpdateConfig(ctx, r.eventDispatcherConfigStore.GetConfig(), config)
-	if err != nil {
-		logging.FromContext(ctx).Error("Error updating InMemory dispatcher config")
+		logging.FromContext(ctx).Error("Error creating config for in memory channels", zap.Error(err))
 		return err
 	}
 
+	// First grab the MultiChannelFanoutMessage handler
+	handler := r.multiChannelMessageHandler.GetChannelHandler(config.HostName)
+	if handler == nil {
+		// No handler yet, create one.
+		fanoutHandler, err := fanout.NewFanoutMessageHandler(logging.FromContext(ctx).Desugar(), channel.NewMessageDispatcher(logging.FromContext(ctx).Desugar()), config.FanoutConfig, r.reporter)
+		if err != nil {
+			logging.FromContext(ctx).Error("Failed to create a new fanout.MessageHandler", err)
+			return err
+		}
+		r.multiChannelMessageHandler.SetChannelHandler(config.HostName, fanoutHandler)
+	} else {
+		// Just update the config if necessary.
+		haveSubs := handler.GetSubscriptions(ctx)
+		if diff := cmp.Diff(config.FanoutConfig.Subscriptions, haveSubs, cmpopts.IgnoreFields(kncloudevents.RetryConfig{}, "Backoff", "CheckRetry")); diff != "" {
```

Comment:
Why do you ignore `Backoff` and `CheckRetry`? If these are modified, i expect the subs to be modified

Reply:
That was actually the previous behaviour as well. But your point is valid, but because these are function pointers, we can't really compare them here. We need to compare them properly ( I think we might want to stash the values and diff that instead of diffing the functions):
https://github.com/knative/eventing/pull/4359/files/ad09ca4331f56ff54a8fad47bdd0c3d0ba0bfc9f#diff-52794e684ba20cecd3c453db5ffdd9642cd363f7157ea5566c84e5a755dc5dcaL86

Reply:
Fair point, probably more than storing closures we should store concrete struct with the related logic. Can you open an issue for this one?

Reply:
https://github.com/knative/eventing/issues/4375